

# 5 Numpy functions to ease out your number work!


### Mathematical Functions in Numpy

Numpy is an numerical python library that is used to work with multidimensional arrays and matrices.Numpy provides a huge variety of functions that help us in exploring data.

- 1)numpy.diff()
- 2)numpy.clip()
- 3)numpy.nan_to_num()
- 4)numpy.fix()
- 5)numpy.cumsum()


In [16]:
!pip install jovian --upgrade -q

In [17]:
import jovian

In [18]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Updating notebook "lakshitagaur123/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/lakshitagaur123/numpy-array-operations


'https://jovian.ai/lakshitagaur123/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [5]:
import numpy as np
from numpy.random import randn,randint

In [8]:
# List of functions explained 
function1 = np.diff
function2 = np.clip
function3 = np.nan_to_num
function4 = np.fix
function5 = np.cumsum

## Function 1 - np.diff()

This numpy function returns the nth difference of consecutive elements of an array. We can input the number of consective elements of which we need to find the difference. 
For eg-: 
- If we need to find the difference between pairs we have n=1 (a[i+1] - a[i])
- If we need to find the differnce between pairs of the array we got as a result for n=1... we define n as n=2 and the number keeps on increasing with the number of differntiated arrays. (a[i+1] - a[i])-(a[i+3] - a[i+2])....
- n basically represents the number of times subtraction operations need to be performed between the consecutive elements


np.diff(a, n=?,axis =?, prepend=<value>,append=<value>)

In [10]:
# Example 1
#Creation of an array
arr= np.random.randint(1,100,10)
arr

array([21, 27, 95, 89, 46,  1, 86, 97, 92, 26])

In [14]:
#Using diff function
b= np.diff(arr,n=1,axis=0)
b

array([  6,  68,  -6, -43, -45,  85,  11,  -5, -66])

In the above example n=1, arr and axis are passed as arguments. As n=1 consecutive elements will get subtracted and the difference will be stored in another array, here it is 'b'. 
If we look at the example-:
When n=1 ,  a[i+1]- a[i] 27-21 = 6 
,95-27 = 68 and so on ... 

In [17]:
# Example 2 - working
arr1= np.random.randint(50,70,11)
arr1

array([62, 64, 54, 53, 59, 56, 64, 61, 63, 60, 67])

In [21]:
#Using diff function
c= np.diff(arr1)
d= np.diff(arr1,n=2)
print(f"With n =1 {c}")
print(f"With n =2 {d}")

With n =1 [  2 -10  -1   6  -3   8  -3   2  -3   7]
With n =2 [-12   9   7  -9  11 -11   5  -5  10]


We already know about the array we got as n=1, n=2 gives us the difference from the array output of n=1, 
Thus array d has = -10-2 = -12 ,  -1-(-10)= 9 and so on...

In [34]:
# Example 3 - breaking (to illustrate when it breaks)
arr= np.array([1,2,3,'eye','uhy',0,100,78])
arr

array(['1', '2', '3', 'eye', 'uhy', '0', '100', '78'], dtype='<U21')

In [35]:
np.diff(arr,n=1)

UFuncTypeError: ufunc 'subtract' did not contain a loop with signature matching types (dtype('<U21'), dtype('<U21')) -> None

In this example I took an array with int and string values.. value mismatched and subtraction could not take place, we should always keep in mind to take values of same type.

#### In our real lives this example can be easily used in ... cracking pattern based questions in aptitude exams.. this function can even be used in finding the nth discrete values

In [36]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "lakshitagaur123/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/lakshitagaur123/numpy-array-operations


'https://jovian.ai/lakshitagaur123/numpy-array-operations'

## Function 2 - np.clip()

The clip function in python gives us the value in the range or interval  which are asked by the user.
- For Example - : IF the user gives min value to be 4 and max value to be 80 in an array... 
 - the clip function will modify the values array element<4 to be equal to 4 
 - array element > 80 to be equal to 80.

In [39]:
#Example 1 - working

a= np.arange(45,100,3)
print(f"Unclipped Array {a}")
b= np.clip(a,a_min =50,a_max=85)
print(f"Clipped Array {b}")

Unclipped Array [45 48 51 54 57 60 63 66 69 72 75 78 81 84 87 90 93 96 99]
Clipped Array [50 50 51 54 57 60 63 66 69 72 75 78 81 84 85 85 85 85 85]


In the above example.. the Minimum value is given to be 50 and Maximum value is given to be 85.....
so elements less than 50 i.e, 45 and 48 are modified to 50 and elements greater than 85 are modified to 85 i.e, 87,90,93,96 and 99

In [44]:
# Example 2 - working
a = np.array([2,5,1,3,6,7,9,0,4,1])
v= np.clip(a, [3, 3, 3, 3, 4, 9, 1, 2, 5, 8],7)
print(f"Unclipped Array {a}")
print(f"Clipped Array {v}")


Unclipped Array [2 5 1 3 6 7 9 0 4 1]
Clipped Array [3 5 3 3 6 7 7 2 5 7]


In the above example the, in place of giving an integer value to be minimum... I have given an array. All the elements of the array passed in the argument will be compared with the values of the corresponding elements of array given by the user... and after that the values lesser than the compared values are modified...

In [45]:
# Example 3 - breaking (to illustrate when it breaks)

a = np.array([2,5,1,3,6,7,9,0,4,1,11,12])
v= np.clip(a, [3, 3, 3, 3, 4, 9, 1, 2, 5, 8],7)

ValueError: operands could not be broadcast together with shapes (12,) (10,) () 

In this example, I took arrays of different shapes... The unclipped array with shape (12,) and the array passed as min argument with shape (10,)... As the shapes are not same.. All the corresponding values will not be compared and some would be left out.. The arrays wouldnt be broadcasted....
Thus, if you are passing an array as an argument .. remember to keep it in same shape as the inputted array

#### This function can be easily used to clean data which is out of interval defined by the user.

In [32]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/numpy-array-operations


'https://jovian.ml/aakashns/numpy-array-operations'

## Function 3 - np.lcm()

This Function returns the lcm (Least Common Multiple) between two arrays or two inputs given as asrguments x1 and x2.

In [8]:
# Example 1 - working
a= np.arange(1,6)
b=np.arange(11,16)
np.lcm(a,b)

array([11, 12, 39, 28, 15])

In the above example, the output array returns the lcm by comparing the corresponding elements of both the arrays.

In [10]:
# Example 2 - working
np.lcm.reduce([20,70,68,92])

54740

The reduce function.. reduces the dimension of the array by one. That's why lcm function returned the lcm of 4 numbers together.

In [11]:
# Example 3 - breaking (to illustrate when it breaks)
np.lcm([20,70,68,92])

TypeError: lcm() takes from 2 to 3 positional arguments but 1 were given

In this example, reduce operation wasn't used thus the array passed was considered as a single argument and didnt have any other argument to be compared with. That's why the error popped up.

#### This function can easily be used to find the common multiples between a variety of numbers which can be useful to analyze data

In [12]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "lakshitagaur123/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/lakshitagaur123/numpy-array-operations


'https://jovian.ai/lakshitagaur123/numpy-array-operations'

## Function 4 - np.fix()

This functions rounds the value to nearest integer and returns the value in float data type

In [3]:
# Example 1 - working
arr = np.arange(1.5,51.6,3.2)
arr


array([ 1.5,  4.7,  7.9, 11.1, 14.3, 17.5, 20.7, 23.9, 27.1, 30.3, 33.5,
       36.7, 39.9, 43.1, 46.3, 49.5])

In [4]:
c=np.fix(arr)
c

array([ 1.,  4.,  7., 11., 14., 17., 20., 23., 27., 30., 33., 36., 39.,
       43., 46., 49.])

In the above example all the numbers with decimals were converted to the nearest integer number ... For eg., 1.5 -> 1.
and 14.3 -> 14

In [9]:
# Example 2 - working
np.fix(23.196)

23.0

You can easily easily round up a single number too, using the fix function.

In [11]:
# Example 3 - breaking (to illustrate when it breaks)
np.fix(23.5,87.222,90.1)

TypeError: func() takes from 1 to 2 positional arguments but 3 were given

In this example i tried to fix all three decimal numbers together, without passing them as a single argument in the form of an array. Thus the type error showcased

#### In real life, none of us now deals with paisa.. that is we always pay our amount in rupees and ignore the decimal values of paisa... thus this fix function can easliy be used to determine the right amount to be paid.

In [17]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "lakshitagaur123/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/lakshitagaur123/numpy-array-operations


'https://jovian.ai/lakshitagaur123/numpy-array-operations'

## Function 5 - np.cumsum()

This function returns the commulative sum of elements in an array and stores them..
For eg., Input array is [1,2,3] after cumsum function = [1,3,6]. The elements keep on adding 1 comes first .. then 3 because 1+2 =3 and then 6 because 1+2+3 = 6

In [21]:
# Example 1 - working
arr = np.arange(1,20,2)
arr


array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

In [22]:
c=arr.cumsum()
c

array([  1,   4,   9,  16,  25,  36,  49,  64,  81, 100])

In this example the output array 'C' after cumsum function is performed returns commulative sum as .... 9 is returned after the addition of 1, 3 and 5 .....36 is returned after the addition of 1,3,5,7,9,11.

In [24]:
# Example 2 - working
arr = np.arange(1,51,2).reshape(5,5)
arr

array([[ 1,  3,  5,  7,  9],
       [11, 13, 15, 17, 19],
       [21, 23, 25, 27, 29],
       [31, 33, 35, 37, 39],
       [41, 43, 45, 47, 49]])

In [29]:
d= np.cumsum(arr, axis=0,dtype='float')
d

array([[  1.,   3.,   5.,   7.,   9.],
       [ 12.,  16.,  20.,  24.,  28.],
       [ 33.,  39.,  45.,  51.,  57.],
       [ 64.,  72.,  80.,  88.,  96.],
       [105., 115., 125., 135., 145.]])

In this example, I have taken the axis to be 0 in this multidimensional array, so the cumsum comes out to be horizontally...that is row wise .. and the dtype is given to be Float .. which means the user wants the output array to be in float data type

In [33]:
# Example 3 - breaking (to illustrate when it breaks)
b= np.arange(1,26)
d= np.cumsum(arr, axis=0,dtype='float',out=b)


ValueError: Iterator input op_axes[0][0] (==1) is not a valid axis of op[0], which has 1 dimensions 

Out is used to define the array in which we want our output to be stored... but in this case i have defined b as a 1D array and d is 2D array. Thus, as the shapes of the arrays are different it wont be compatible......... Always remember that the array declared in out should be of the same shape

#### This Function can be easliy used in finding the sum of elements from a particular time..If a dataset of prices is given  with 12 months and we need to find the price till the 4th month. We can use the cumsum operation and display the 4th element in the output array.

In [13]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "lakshitagaur123/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/lakshitagaur123/numpy-array-operations


'https://jovian.ai/lakshitagaur123/numpy-array-operations'

## Conclusion

### Numpy Functions prove to be great when they are used to analyze a given piece of data. Sometimes, the client requires us to perform some of the functions which can focus on real life mathematical problems as well. Here, I made you familiar with 5 basic mathematical numpy functions which can be useful for your work ahead. For more such functions you can have a look at the documentation. Links are provided below

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html


In [19]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "lakshitagaur123/numpy-array-operations" on https://jovian.ai
[jovian] Committed successfully! https://jovian.ai/lakshitagaur123/numpy-array-operations


'https://jovian.ai/lakshitagaur123/numpy-array-operations'

### - Lakshita Gaur